Spigot-Server-Patches/0420-mplement-Mob-Goal-API.patch

```diff
@@ -0,0 +1,662 @@
+@@ -0,0 +1,46 @@
++package com.destroystokyo.paper.ai;
++
++import net.minecraft.server.PathfinderGoal;
++
++/**
++ * Wraps api in vanilla
++ */
++public class PaperCustomGoal extends PathfinderGoal {
++
++    private Goal handle;
++
++    public PaperCustomGoal(Goal handle) {
++        this.handle = handle;
++
++        this.setFlags(MobGoalHelper.paperToVanilla(handle.getFlags()));
++    }
++
++    @Override
++    public boolean shouldActivate() {
++        return handle.shouldActivate();
++    }
++
++    @Override
++    public boolean shouldStayActive() {
++        return handle.shouldStayActive();
++    }
++
++    @Override
++    public void start() {
++        handle.start();
++    }
++
++    @Override
++    public void onTaskReset() {
++        handle.stop();
++    }
++
++    @Override
++    public void tick() {
++        handle.tick();
++    }
++
++    public Goal getHandle() {
++        return handle;
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/ai/PaperMobGoals.java b/src/main/java/com/destroystokyo/paper/ai/PaperMobGoals.java
+new file mode 100644
+index 000000000..07ca533f6
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/ai/PaperMobGoals.java
+@@ -0,0 +1,116 @@
++package com.destroystokyo.paper.ai;
++
++import net.minecraft.server.PathfinderGoal;
++import net.minecraft.server.PathfinderGoalSelector;
++import net.minecraft.server.PathfinderGoalWrapped;
++
++import java.util.Collection;
++import java.util.EnumMap;
++import java.util.HashMap;
++import java.util.HashSet;
++import java.util.Map;
++import java.util.Set;
++
++import org.bukkit.NamespacedKey;
++
++public class PaperMobGoals implements MobGoals {
++
++    private Map<PathfinderGoal, PaperVanillaGoal> instanceCache = new HashMap<>();
++
++    private EnumMap<GoalType, PathfinderGoalSelector> handles;
++
++    public PaperMobGoals(EnumMap<GoalType, PathfinderGoalSelector> handles) {
++        this.handles = handles;
++    }
++
++    @Override
++    public void addGoal(GoalType type, int priority, Goal goal) {
++        checkType(type);
++        getHandle(type).addGoal(priority, new PaperCustomGoal(goal));
++    }
++
++    @Override
++    public void removeGoal(GoalType type, Goal goal) {
++        checkType(type);
++        removeGoal(type, goal.getKey());
++    }
++
++    @Override
++    public void removeAllGoals() {
++        for (GoalType type : handles.keySet()) {
++            removeAllGoals(type);
++        }
++    }
++
++    @Override
++    public void removeAllGoals(GoalType type) {
++        checkType(type);
++        for (Goal goal : getAllGoals(type)) {
++            removeGoal(type, goal);
++        }
++    }
++
++    @Override
++    public void removeGoal(GoalType type, NamespacedKey key) {
++        checkType(type);
++        for (Goal g : getAllGoals(type)) {
++            if (g.getKey().equals(key)) {
++                getHandle(type).removeGoal(((PaperVanillaGoal) g).getHandle());
```

Comment:
This API only allows removing Vanilla goals? Once I've add my own custom goal with this API, I can't remove it?
```
java.lang.ClassCastException: class com.theroadmc.bossfight.TeleportBehindGoal cannot be cast to class com.destroystokyo.paper.ai.PaperVanillaGoal (com.theroadmc.bossfight.TeleportBehindGoal is in unnamed module of loader org.bukkit.plugin.java.PluginClassLoader @4b9ddd56; com.destroystokyo.paper.ai.PaperVanillaGoal is in unnamed module of loader 'app')
	at com.destroystokyo.paper.ai.PaperMobGoals.removeGoal(PaperMobGoals.java:58) ~[paper.jar:git-Paper-"1141ed4"]
...
```

Reply:
will look into that, thanks

Reply:
fixed